pkg/controllermanager/controller/cloudprofile/kubernetesversion/kubernetes_version.go

```diff
@@ -0,0 +1,164 @@
+// returns true if the CloudProfile has to be updated, the updated CloudProfile or an error
+func ReconcileKubernetesVersions(logger *logrus.Entry, config *controllermgrconfig.CloudProfileControllerConfiguration, cloudProfile *gardencorev1beta1.CloudProfile) (bool, *gardencorev1beta1.CloudProfile, error) {
+	if config == nil || config.KubernetesVersionManagement == nil || config.KubernetesVersionManagement.Enabled == false {
+		return false, cloudProfile, nil
+	}
+
+	profile := cloudProfile.DeepCopy()
+	supportedVersionsMap := make(map[string]helper.HighestSupportedVersion, len(profile.Spec.Kubernetes.Versions))
+	for cloudProfileIndex, version := range profile.Spec.Kubernetes.Versions {
+		currentSemverVersion, err := semver.NewVersion(version.Version)
+		if err != nil {
+			return false, nil, err
+		}
+
+		hasSupportedMinorVersion, err := gardencorev1beta1helper.HasSupportedKubernetesMinorVersion(*config.KubernetesVersionManagement.SupportedKubernetesVersions, profile, *currentSemverVersion)
+		if err != nil {
+			return false, nil, err
+		}
+
+		newerPatchVersionFound, _, err := gardencorev1beta1helper.DetermineLatestKubernetesPatchVersion(profile, version.Version)
+		if err != nil {
+			return false, nil, err
+		}
+		// better readability
+		isLatestPatchVersion := !newerPatchVersionFound
+
+		// TODO: Remove migration code in future version
+		// assumes there is no preview version yet - otherwise no version will be set to supported
+		if version.Classification == nil && isLatestPatchVersion && hasSupportedMinorVersion {
+			// highest patch version of supported minor version -> supported
+			logger.Infof("[KUBERNETES VERSION MANAGEMENT] setting '%s' classification for kubernetes version '%s' that had no classification", gardencorev1beta1.ClassificationSupported, version.Version)
+			version = helper.SetSupportedClassification(version)
+			profile.Spec.Kubernetes.Versions[cloudProfileIndex] = version
+		} else if version.Classification == nil {
+			// deprecate
+			deprecatedVersion, err := deprecateKubernetesVersion(logger, config, profile, version, *currentSemverVersion, hasSupportedMinorVersion, isLatestPatchVersion)
+			if err != nil {
+				return false, nil, err
+			}
+			profile.Spec.Kubernetes.Versions[cloudProfileIndex] = *deprecatedVersion
+			continue
+		} else if !hasSupportedMinorVersion && (*version.Classification == gardencorev1beta1.ClassificationPreview || *version.Classification == gardencorev1beta1.ClassificationSupported) {
+			deprecatedVersion, err := deprecateKubernetesVersion(logger, config, profile, version, *currentSemverVersion, false, isLatestPatchVersion)
+			if err != nil {
+				return false, nil, err
+			}
+			profile.Spec.Kubernetes.Versions[cloudProfileIndex] = *deprecatedVersion
+			continue
+		}
+
+		// only keep one supported version
+		if version.Classification != nil && *version.Classification == gardencorev1beta1.ClassificationSupported {
```

Comment:
Isn't it possible to check that the cloud profile only contains one "supported" version classifier per major/minor K8s by a validation function?

Reply:
Yes it would be, however then the deprecating of the currently supported version with the right deprecation times has to be done manually. Thats why it is automated here. I also added that in the docs.